src/controller/CHIPDevice.cpp

```diff
@@ -88,11 +99,16 @@ CHIP_ERROR Device::SendMessage(System::PacketBufferHandle buffer, PayloadHeader
 
         ReturnErrorOnFailure(LoadSecureSessionParameters(ResetTransport::kYes));
 
-        err = mSessionManager->SendMessage(mSecureSession, std::move(resend));
+        err = exchange->SendMessage(protocolId, msgType, std::move(resend), sendFlags);
         ChipLogDetail(Controller, "Re-SendMessage returned %d", err);
         ReturnErrorOnFailure(err);
     }
 
+    if (exchange != nullptr)
+    {
+        exchange->Close();
```

Comment:
If we reach the resend codepath and it hits a failure for any of its `ReturnErrorOnFailure` calls, we will skip doing this step and end up leaking an exchange, right?

I'm OK with a followup to track this, just so we don't keep blocking this PR, but please file an issue to address this (e.g. by adding some sort of RAII exchange-closing class).

Reply:
Filed https://github.com/project-chip/connectedhomeip/issues/5922